File: scripts/Contains Duplicate III/Contains Duplicate III.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number} t
 * @return {boolean}
 */
var containsNearbyAlmostDuplicate = function(nums, k, t) {
    if(nums.length <2) return false;
    
    let isValid = false;
    
    for(var indexI=0; indexI<nums.length; indexI++){
       let indexJ =  indexI+1;
       while((indexJ - indexI) <= k){
           if(Math.abs(nums[indexI] - nums[indexJ]) <= t){
              isValid= true;
              break;
           }
           indexJ++;
       }
      if (isValid) break;
    }
    
    return isValid;
};

File: algorithms/C/Contains Duplicate III/Contains Duplicate III.java
class Solution {
public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
    TreeSet<Long> treeSet = new TreeSet<>();
    //nums = [1,2,3,1], k = 3, t = 0
    for(int i = 0; i < nums.length; i++){
        Long cur = (long) nums[i];
        //greatest element that is less than or equal to current
        Long floor = treeSet.floor(cur);
        //smallest element that is greater than or equal to current
        Long ceiling = treeSet.ceiling(cur);
        
        if(null != floor && Math.abs(floor - cur) <= t)
            return true;
        
        if(null != ceiling && Math.abs(ceiling - cur) <= t)
            return true;
        
        treeSet.add(cur);
        if(treeSet.size() > k)
            treeSet.remove((long)nums[i-k]);            
    }
    return false;        
}
}